pkg/reconciler/pipelinerun/pipelinerun_test.go

```diff
@@ -3753,3 +3456,40 @@ func getTaskRunStatus(t string, status corev1.ConditionStatus) *v1beta1.Pipeline
 		},
 	}
 }
+
+func (prt PipelineRunTest) reconcileRun() (*v1beta1.PipelineRun, test.Clients) {
+	prt.Test.Helper()
+	testAssets, cancel := getPipelineRunController(prt.Test, prt.Data)
+	defer cancel()
+	c := testAssets.Controller
+	clients := testAssets.Clients
+	recorder := testAssets.Recorder
+
+	reconcileError := c.Reconciler.Reconcile(context.Background(), prt.Namespace+"/"+prt.PRName)
+	if prt.PermanentError {
+		// When a PipelineRun is invalid and can't run, we expect a permanent error that will
+		// tell the Reconciler to not keep trying to reconcile.
+		if reconcileError == nil {
+			prt.Test.Fatalf("Expected an error to be returned by Reconcile, got nil instead")
+		}
+		if controller.IsPermanentError(reconcileError) != prt.PermanentError {
+			prt.Test.Fatalf("Expected the error to be permanent: %v but got: %v", prt.PermanentError, controller.IsPermanentError(reconcileError))
+		}
+	} else if reconcileError != nil {
+		prt.Test.Fatalf("Error reconciling: %s", reconcileError)
+	}
+	// Check that the PipelineRun was reconciled correctly
+	reconciledRun, err := clients.Pipeline.TektonV1beta1().PipelineRuns(prt.Namespace).Get(prt.PRName, metav1.GetOptions{})
+	if err != nil {
+		prt.Test.Fatalf("Somehow had error getting reconciled run out of fake client: %s", err)
+	}
+
+	// Check generated events match what's expected
+	if len(prt.WantEvents) > 0 {
```

Comment:
this is fairly minor but something feels weird to me about having WantEvents and PRName be attributes of PipelineRunTest - i think it's that this makes it so that if you want to call reconcileRun more than once in the same test, e.g. for two different pipelineruns, one reconciled after the other, you'd need to modify PipelineRunTest

I could see making 2 changes:
- Make WantEvents and PRName and Namespace arguments to this function
-  Have a NewPipelineRunTest function that returns an instance of PipelineRunTest and takes Test and Data as arguments

this way you can call NewPipelineRunTest to create your controller with the state it needs (store the testAssets as an attribute of PipelineRunTest) and you can call reconcileRun with as many runs as you want

Reply:
Hey @bobcatfish, I have added [NewPipelineRunTest](https://github.com/tektoncd/pipeline/pull/2876/files#diff-f7c8532cfe79319c0fff711a4979a6d6R3633) and updated [reconcileRun](https://github.com/tektoncd/pipeline/pull/2876/files#diff-f7c8532cfe79319c0fff711a4979a6d6R3644) by adding function parameters (Namespace, PipelineRunName, wantEvents, and permanentError). 

Now, the PipelineRunTest has:

```
        test.Data  `json:"inline"`
	Test       *testing.T
	TestAssets test.Assets
	Cancel     func()
```

Also, have added `defer` on `Cancel` as part of each unit test instead of making it part of common function.

Reply:
`Data` can be dropped from the `PipelineRunTest` struct since  `NewPipelineRunTest` is creating the controller and  instantiating `PipelineRunTest` with assets, `data` is not used/needed anywhere beyond `NewPipelineRunTest` - It makes `PipelineRunTest` complete (with inputs and assets) but can create separate PR to drop it if needed 😉